WebUI: Support managing category download path

Piccirello opened the pull request:
![Screenshot 2025-06-27 at 22 46 22](https://github.com/user-attachments/assets/c193cb5a-c784-4ab1-b42d-2f692271ed5a)

glassez commented on src/base/bittorrent/categoryoptions.cpp:
```diff
@@ -52,7 +52,7 @@ BitTorrent::CategoryOptions BitTorrent::CategoryOptions::fromJSON(const QJsonObj
 QJsonObject BitTorrent::CategoryOptions::toJSON() const
 {
-    QJsonValue downloadPathValue = QJsonValue::Undefined;
+    QJsonValue downloadPathValue = QJsonValue::Null;
```
And if this change is considered to be bugfix I would extract it into separate PR.

Piccirello replied:
Yes, I just figured since the conversation is happening here, I would let it complete here before moving the code and likely having to change it.

Piccirello replied:
Created #22958

glassez replied:
> Created #22958

Changes related to updating the WebAPI version should also be moved there, right?

Piccirello commented on src/webui/www/private/newcategory.html:
```diff
@@ -146,17 +207,43 @@
                 }
             });
 
+            document.getElementById("useDownloadPath").addEventListener("change", (e) => {
```
I tried to consolidate these when writing this PR, but I've just consolidated a bunch more 👍

Piccirello commented:
> When "Use another path for incomplete torrents" in qBt settings is checked and the corresponding choice in this dialog is
> 
> 1. "Default": It should fill the (disabled) field with the incomplete save path that will be used (now it's empty)

Fixed! Another good catch.

> 2. "Yes": This is just a suggestion because currently it follows the GUI behavior. When the incomplete save path has not been overwritten here, IMO it should also fill the field with actual text and not just placeholder text with the incomplete save path that will be used. Why? I think it's easier to change a part of the path than to write the whole thing. I doubt that people type drastically different save paths often where the current way is more convenient.

I agree with this. I've switched the behavior.

> Note that it now allows you to override whether or not an incomplete path is used without specifying a path itself (i.e. use default incomplete path).

This is true, though I think it's not intuitive for users. They may not expect that leaving the field blank will use the default path. At first I thought it was a bug that the GUI allowed using a blank path, until I looked at the code.

glassez commented:
> This is true, though I think it's not intuitive for users. They may not expect that leaving the field blank will use the default path.

If so, wouldn't it be better to simply add some explanation label next to it rather than making this feature inaccessible?

Piccirello commented:
> > This is true, though I think it's not intuitive for users. They may not expect that leaving the field blank will use the default path.
> 
> If so, wouldn't it be better to simply add some explanation label next to it rather than making this feature inaccessible?

The feature is still accessible, the user just has to delete the existing pre-filled value before submitting the form.

glassez commented:
> The feature is still accessible, the user just has to delete the existing pre-filled value before submitting the form.

IMO, this is less intuitive than the opposite action (fill in the field if you want to explicitly set the path).

Piccirello commented:
> > The feature is still accessible, the user just has to delete the existing pre-filled value before submitting the form.
> 
> IMO, this is less intuitive than the opposite action (fill in the field if you want to explicitly set the path).

Both are pretty unintuitive. For now I'll keep the WebUI's behavior matching the GUI and then someone can improve both if they'd like.

thalieht commented:
> > I think it's easier to change a part of the path than to write the whole thing. I doubt that people type drastically different save paths often
> 
> IMO, in most cases, the default path is sufficient for the user, which has the format `<base_path>/<category_name>`. However, we provide the option to enable or disable it for specific categories.

I don't understand what you said is related.
Anyway i hadn't thoroughly thought that when the field is filled it means that the user changed the path explicitly.

Piccirello commented:
Rebased onto master now that #22932 is merged. This PR now only contains the changes to the WebUI.

glassez commented:
@Piccirello 
WebUI hasn't full featured "Watched folder options" editor. So you could take care of it if you want.

Chocobo1 commented on src/webui/www/private/newcategory.html:
```diff
@@ -66,6 +117,16 @@
                     return true;
                 };
 
+                const body = {};
```
Not sure if this is worth it but you can use a `Map` instead of a `Object` as such:
```js
const body = new Map();
body.set("downloadPathEnabled", false);

const params = {
    ...Object.fromEntries(body),
}
```

Chocobo1 commented on src/webui/www/private/newcategory.html:
```diff
@@ -74,6 +135,7 @@
                         fetch("api/v2/torrents/createCategory", {
                                 method: "POST",
                                 body: new URLSearchParams({
+                                    ...body,
```
I would suggest renaming the variable to avoid this little confusing pattern:
```suggestion
                                body: new URLSearchParams({
                                    ...dlPathParams,
```

Chocobo1 commented:
@Piccirello 
Thank you!
